### ISSUE TITLE
Change default behavior of scaled_dot_product_attention's causal masking alignment

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#108108]
## Current Behavior:
The `torch.nn.functional.scaled_dot_product_attention` function currently applies causal masking to the top left corner of the attention matrix. This well defined when seqlen_q equals seqlen_kv, but an alignment choice needs to made when they are not.

#### Current alignment choice
If seqlen_q != seqlen_kv and causal=True, the causal mask is aligned to the the top-left corner.

For example, if seqlen_q = 2 and seqlen_kv = 5, the causal mask (1 = keep, 0 = masked out) is:
```
    1 0 0 0 0
    1 1 0 0 0
```

If seqlen_q = 5 and seqlen_kv = 2, the causal mask is:
```
    1 0
    1 1
    1 1
    1 1
    1 1
```

## Proposal:
I propose changing the default behavior of the scaled_dot_product_attention function's causal masking to align with the bottom right corner of the attention matrix. This change would be backward incompatible for inputs when seqlen_q does not equal seqlen_kv, as it would shift the masking pattern.

For example, if seqlen_q = 2 and seqlen_kv = 5, the causal mask (1 = keep, 0 = masked out) is:
```
    1 1 1 1 0
    1 1 1 1 1
```
If seqlen_q = 5 and seqlen_kv =  2, the causal mask is:
```
    0 0
    0 0
    0 0
    1 0
    1 1
 ```
_If the row of the mask is all zero, the output will be zero._

## Benefits:

This choice of mask is beneficial when performing autoregressive decoding for LLMs. 
For example lets say that your prefill prompt consisted of 4 tokens and you had a max KV cache size of 8. Then at step 1 of decoding, your query sequence length would be size 1 (index position 5) and your KV input would be of size 5. The query's seqlen is 1 but it really represents the 5 token in the sequence. So the mask for this attention should ideally look:
```
1 1 1 1 1 0 0 0
```

With the existing implementation of causal, the produced attention mask would be. 
```
1 0 0 0 0 0 0 0 
``` 

## Drawbacks:

Backward compatibility: The proposed change would break backward compatibility for cases where seqlen_q doesn't equal seqlen_kv.
Existing code: Users who rely on the current masking alignment would need to update their code if they want to adopt the new behavior.

## Implementation:
To implement this change, the following steps need to be taken:

- Modify the `scaled_dot_product_attention_math` function to align the causal masking with the bottom right corner of the attention matrix.
- Change the default call `_scaled_dot_product_efficient_attention` to https://github.com/pytorch/pytorch/blob/3488837ec118cc8f8bf9fe10a8c5a32726ba0a7b/aten/src/ATen/native/transformers/cuda/attention.cu#L734 to use the bottom right
- After updating to FlashAttentionV2 we would apply the changes from this commit: https://github.com/Dao-AILab/flash-attention/commit/9e5e8bc91e30af5cdc321362b553f6c0da332e30 
- Update the fused CPU implementation to obey the above semantics.
- Update the function's documentation and any relevant examples to reflect the new masking behavior.

## Additional Context:
Both Xformers and FlashAttention have made this behavior the new default

**Comment by ydshieh** [comment]
Hi @drisspg . Is/would this proposal under working in progress? I checked with torch 2.4 (RC) and torch 2.5 (nightly) and it looks like they are still using the current mask (as in torch 2.3).

I also have a question about the benefit mentioned

> For example lets say that your prefill prompt consisted of 4 tokens and you had a max KV cache size of 8. Then at step 1 of decoding, your query sequence length would be size 1 (index position 5) and your KV input would be of size 5. The query's seqlen is 1 but it really represents the 5 token in the sequence. So the mask for this attention should be:


In this case, query sequence length would be size 1, and KV (cache) length is of size 8, statically allocated. Now even if with the proposal (that is the same with current flash attention >= 2.1), the attention mask would be 

> 1 1 1 1 1 1 1 1

and the query still attends to the last 3 tokens (position index  6, 7, 8 , 1-based indexing).

I think this has to be addressed in flash attention library too in order to use static KV cache. Let me know if I make some mistake here, thank you.

**Comment by ydshieh** [comment]
Flash attention now have `flash_attn_with_kvcache` that could deal with the training 0 in the mask by using the argument `cache_seqlens` + the length of new key/value.

However, if pytorch ever want to update sdpa with FA2, it's likely we will have a `scaled_dot_product_attention_with_kvcache` rather adding extra arguments into the existing `flash_attn_with_kvcache` ?

**Comment by ydshieh** [comment]
OK, I see in #111007

> // for non-square masks. We will not support non-square masks for causal w/ FAV2

and #110681.

Not sure if it's still WIP.

**Comment by drisspg** [comment]
@ydshieh  You can find the mechanism for using lower-right-causal masking here: https://github.com/pytorch/pytorch/blob/f86dbae2471bcfddd68d11b8baa3c952fcf8b2e9/torch/nn/attention/bias.py#L321

You are also right that for static kv_cache we would need a different mechanism for supporting this. We currently dont support sdpa_with_kv_cache. The existing plan has been to do something using NestedTensor 

That being said we still might  ship a kv_cache enabled variant of sdpa

**Comment by lxt160980** [comment]
Hi, sorry for bothering ... In your examples you mentioned :
If seqlen_q = 5 and seqlen_kv = 2, the causal mask is:
```
    0 0
    0 0
    0 0
    1 0
    1 1
```
I couldn't understand how this works. Because for me I suppose the query is the one who gets generated & updated, say we want to generate 2 new tokens at once,  in this case, the newest causal mask  should be :
```
    1 1
    1 1
```
instead of 
```
    1 0
    1 1
```
What is the mapping relationship between  seqlen_q seqlen_kv and what scenario is this? I am struggling with the basic background info, maybe I get the context wrong.
Hoping for your reply!
